src/test/java/GoServletTest.java

```diff
@@ -0,0 +1,254 @@
+import com.google.maps.model.DirectionsLeg;
+import com.google.maps.model.DirectionsResult;
+import com.google.maps.model.DirectionsRoute;
+import com.google.maps.model.Duration;
+import com.google.maps.model.GeocodingResult;
+import com.google.maps.model.Geometry;
+import com.google.maps.model.LatLng;
+import com.google.maps.model.PlaceType;
+import com.google.maps.model.RankBy;
+import com.google.sps.model.DirectionsClient;
+import com.google.sps.model.DirectionsClientFactory;
+import com.google.sps.model.GeocodingClient;
+import com.google.sps.model.GeocodingClientFactory;
+import com.google.sps.model.PlacesClient;
+import com.google.sps.model.PlacesClientFactory;
+import com.google.sps.model.TasksClient;
+import com.google.sps.model.TasksClientFactory;
+import com.google.sps.servlets.GoServlet;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mockito;
+
+@RunWith(JUnit4.class)
+public class GoServletTest extends AuthenticatedServletTestBase {
+
+  private DirectionsClientFactory directionsClientFactory;
+  private DirectionsClient directionsClient;
+  private PlacesClientFactory placesClientFactory;
+  private PlacesClient placesClient;
+  private TasksClientFactory tasksClientFactory;
+  private TasksClient tasksClient;
+  private GeocodingClientFactory geocodingClientFactory;
+  private GeocodingClient geocodingClient;
+  private GoServlet servlet;
+
+  private static final String API_KEY = "fake api key";
+  private static final String RESTAURANT_ONE = "restaurant near coordinate one";
+  private static final String RESTAURANT_TWO = "restaurant near coordinate two";
+  private static final String BANK_ONE = "bank near coordinate one";
+  private static final String BANK_TWO = "bank near coordinate two";
+
+  private GeocodingResult streetAddressGeocodingResult = new GeocodingResult();
+  private Geometry streetAddressGeometry = new Geometry();
+  private LatLng streetAddressCoordinates = new LatLng(0, 0);
+  private GeocodingResult restaurantGeocodingResult = new GeocodingResult();
+
+  private Duration longerDuration = new Duration();
+  private DirectionsLeg longerLeg = new DirectionsLeg();
+  private DirectionsRoute longerRoute = new DirectionsRoute();
+  private DirectionsResult longerResult = new DirectionsResult();
+
+  private Duration shorterDuration = new Duration();
+  private DirectionsLeg shorterLeg = new DirectionsLeg();
+  private DirectionsRoute shorterRoute = new DirectionsRoute();
+  private DirectionsResult shorterResult = new DirectionsResult();
+
+  @Override
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+    directionsClientFactory = Mockito.mock(DirectionsClientFactory.class);
+    directionsClient = Mockito.mock(DirectionsClient.class);
+    placesClientFactory = Mockito.mock(PlacesClientFactory.class);
```

Comment:
You shouldn't have to mock a factory

Reply:
Oh true, thanks for pointing that out!

Reply:
Not sure if I am understanding correctly again, kinda need it to give explicit implementations to the GoServlet constructor. The default constructor gets API keys using file I/O.